File: 3383-taking-maximum-energy-from-the-mystic-dungeon/taking-maximum-energy-from-the-mystic-dungeon.java
class Solution {
    public int maximumEnergy(int[] energy, int k) {
        List<Integer> memo = new ArrayList<>(Collections.nCopies(k, 0));
        for (int x : energy)
            memo.add(Math.max(x + memo.get(memo.size() - k), x));

        return Collections.max(memo.subList(memo.size() - k, memo.size()));
    }
}

File: 3383-taking-maximum-energy-from-the-mystic-dungeon/taking-maximum-energy-from-the-mystic-dungeon.go
func maximumEnergy(energy []int, k int) int {
	res := math.MinInt

	for o := range k {
		cur := 0

		for i := o; i < len(energy); i += k {
			cur = max(cur, 0) + energy[i]
		}

		res = max(res, cur)
	}

	return res
}